Add frontend module in launch, launch_xml and launch_yaml packages

ivanpauno opened the pull request:
See https://github.com/ros2/design/pull/207 and https://github.com/ros2/design/pull/208.
Targets all the tasks of https://github.com/ros2/launch/issues/163#issuecomment-464763427 (except the last three).

This is already functional.

TODOs (in importance order):
- Write launch_yaml readme :heavy_check_mark:
- Write more action and substitution parsing functions. :heavy_check_mark:
- Write more xml and yaml examples to exercise the written code. :heavy_check_mark:
- launch_frontend package should probably be merged with launch, as they are completely coupled.
 Action and Substitution parsing methods should be moved to the classes (see 
 https://github.com/ros2/launch_ros/pull/23) (nice to have). :heavy_check_mark:
- Create a `TestEntity` to unit test the parsing methods (nice to have).
- See if we can write a decorator that infers the parsing function from the constructor typing annotations.
 That is a really tricky tasks (nice to have).

P.S.:
`launch_frontend/launch_frontend/parse_substitution.py` (later modified by me) and `launch_frontend/launch_frontend/grammar.lark` were contributed by @hidmic.

Depends on https://github.com/ros2/launch/pull/235 (after https://github.com/ros2/launch/pull/226/commits/3c1c1e9d349e8b1f1c8fd19a0c8e2bb56c0f72a9) (already merged).

hidmic commented on launch_xml/README.md:
```diff
@@ -0,0 +1,67 @@
+
+This package provides an abstraction of the XML tree.
+
+## XML front-end mapping rules
```
@ivanpauno we should also document the attribute lookup behavior (first check attributes in tags, then check attributes as children, then fail).

hidmic commented on launch_frontend/launch_frontend/parser.py:
```diff
@@ -0,0 +1,64 @@
+        args = [args]
+    # TODO(ivanpauno): How will predicates be handle in env?
+    # Substitutions aren't allowing conditions now.
+    env = get_dictionary_from_key_value_pairs(getattr(entity, 'env', None))
```
@ivanpauno why not:
```suggestion
    env = getattr(entity, 'env', None)
    if env is not None:
       env = {e.name: e.value for e in env}
```
?

hidmic commented on launch_frontend/launch_frontend/parser.py:
```diff
@@ -0,0 +1,64 @@
+    prefix = getattr(entity, 'launch-prefix', None)
+    output = getattr(entity, 'output', None)
+    args = getattr(entity, 'args', None)
+    args = args.split(' ') if args else []
```
@ivanpauno so here we have a rather subtle issue. If an attribute isn't found during parsing, we should not be passing `None` to the constructor if it defines a suitable default for that attribute. Same care applies to `str_to_bool` coercion of a `None` variable to a `False` boolean.

hidmic replied:
Still applies, IMHO.

ivanpauno commented on launch_xml/README.md:
```diff
@@ -0,0 +1,67 @@
+`$(list [sep=,] a,b,c,d)`
+: Substituted by a python list, splited by the separator that follows `sep=`.
+  Default separator is `,`.
+  This substitution is evaluated instantaneously, and not in a lazy way.
```
Yes. I think the more general way of supporting list in xml is inferring it from the text. We could have a different notation for this special case.
Another way of handling it is: when accessing the attribute `attr` if the attribute `attr-sep` exists, return `attr` split with `attr-sep` when accessing it.
e.g.:
```xml
<root>
  <tag attr="1,2,3"/>
  <tag2 attr="1,2,3" attr-sep=","/>
</root>
```
```python3
e.tag == '1,2,3'
e.tag2 == ['1', '2', '3']
```
In this last case, we should change the attribute `sep` of `param` to `value-sep` in the xml schema. In my original proposal, that attribute was not more needed.

hidmic replied:
SGTM!

hidmic commented on launch_frontend/launch_frontend/parser.py:
```diff
@@ -0,0 +1,59 @@
+    cmd = entity.cmd
+    cwd = getattr(entity, 'cwd', None)
+    name = getattr(entity, 'name', None)
+    shell = str_to_bool(getattr(entity, 'shell', 'false'))
```
@ivanpauno do we also need other `str_to_*()` functions?

ivanpauno replied:
Probably. When handling node parameters, some kind of automatic conversion of string to int, float and bool will be needed. But it will be an special case for that parsing function.

hidmic commented on launch_frontend/test/launch_frontend/test_substitutions.py:
```diff
@@ -0,0 +1,112 @@
+    subst = default_parse_substitution(
+        r'$(env "asd_bsd_qsd_$(test "asd_bds")" "$(env DEFAULT)_qsd")'
+    )
+    assert len(subst) == 1
```
@ivanpauno missing test?

ivanpauno replied:
Thanks! It was actually a problematic case

hidmic commented on launch_xml/README.md:
````diff
@@ -0,0 +1,95 @@
+When having an xml tag like:
+
+```xml
+<tag attr="2"/>
````
@ivanpauno nit: can we change `attr` to some other string to avoid `get_attr('attr', ...)` below?

hidmic commented on launch_xml/setup.py:
```diff
@@ -0,0 +1,34 @@
+
+setup(
+    name='launch_xml',
+    version='0.7.3',
```
@ivanpauno same about the version.

hidmic commented on launch_xml/test/launch_xml/let_var.xml:
```diff
@@ -0,0 +1,4 @@
+<launch>
+    <let name="var1" value="asd"/>
+    <let name="var2" value="2 $(var var1)"/>
+</launch>
```
@ivanpauno missing new line

hidmic commented on launch_frontend/launch_frontend/parser.py:
```diff
@@ -0,0 +1,111 @@
+        return launch.LaunchDescription(actions)
+
+    @classmethod
+    def load(
```
@ivanpauno meta: for testing, it might make sense to have `load` take a plain string and maybe `load_from` take a file path or object. That way, tests and the description that's being validated can be kept together.

ivanpauno replied:
If I'm understanding well, you want to be able to pass the xml document in an string (and not the file name). Am I right?
Unfortunately, in some cases (like xml), I will have to dump it in a file and read it again ... but that's not much problem.

ivanpauno replied:
I still only have one load method, but I put all the descriptions in place in the tests now.

hidmic commented on launch_yaml/README.md:
````diff
@@ -0,0 +1,103 @@
+```python
+'int', 'float', 'bool', 'list[int]', 'list[float]', 'list[bool]', 'list[str]', 'str'
+```
+
````
@ivanpauno meta: IMHO, type deduction documentation is design document material.

ivanpauno replied:
If you want, I can update https://github.com/ros2/design/pull/208 and https://github.com/ros2/design/pull/207, and create a new PR for yaml design doc.

hidmic replied:
Eventually, yeah, but let's hold that for a minute. At least until we've settled here.

hidmic commented on launch_yaml/launch_yaml/entity.py:
```diff
@@ -0,0 +1,103 @@
+        type_name: Optional[Text] = None,
+        parent: 'Entity' = None
+    ) -> Text:
+        """Construnctor."""
```
@ivanpauno typo `Construnctor` -> `Constructor`

hidmic commented on launch_yaml/launch_yaml/entity.py:
```diff
@@ -0,0 +1,103 @@
+            if len(element) != 1:
+                raise RuntimeError('Expected a len 1 dictionary')
+            self.__type_name = list(element.keys())[0]
+            self.__element = element[self.__type_name]
```
@ivanpauno mind to explain the logic here a bit? Is this to cope with the root dictionary?

ivanpauno replied:
Yeah this is a little confusing ...
- The usual loading is with `type_name=None`. Each yaml launch description is expected to only have one root element, if not an error is raised. Then, I store the (only) key as the name, and the value as the element tree. Having this constructor is also a good idea when using the `children` property, if one of the children is ill formed an error will be raised.
- The "unusual" loading, is with `typename='somestringhere'`. When using `get_attr('sth', types='list[Entity]')`, you have a dict with a list inside. Each of the list elements has to be loaded with the same type_name (the key of the dict).

The real solution would be to always specify the `type_name` argument, and have some formal checking before to check if the yaml file is ill formed or not.

ivanpauno replied:
I finally changed it, enforcing passing `type_name`.

hidmic commented on launch_yaml/setup.py:
```diff
@@ -0,0 +1,34 @@
+
+setup(
+    name='launch_yaml',
+    version='0.7.3',
```
@ivanpauno same thing about the version.

hidmic commented on launch_yaml/test/launch_yaml/test_executable.py:
```diff
@@ -0,0 +1,48 @@
+
+def test_executable():
+    """Parse executable yaml example."""
+    root_entity, parser = Parser.load(str(Path(__file__).parent / 'executable.yaml'))
```
@ivanpauno same comment about parametrizing our tests by the launch description markup as in https://github.com/ros2/launch_ros/pull/23#discussion_r284335310.

ivanpauno replied:
But, where should I put them? Is ok putting them in launch_frontend? And having launch_xml and launch_yaml as test dependencies.
Or maybe, a new package only for the tests.

hidmic replied:
Hmm, good point. Ok, nevermind my comment, it makes it more obscure than what it's worth.

hidmic commented on launch_frontend/launch_frontend/parse_substitution.py:
```diff
@@ -0,0 +1,108 @@
+transformer = ExtractSubstitution()
+
+
+def default_parse_substitution(string_value):
```
Hmm we probably need someone else to review the grammar.

hidmic commented:
> launch_frontend package should probably be merged with launch, as they are completely coupled.
Action and Substitution parsing methods should be moved to the classes (see
ros2/launch_ros#23) (nice to have).

I fully agree and also think we need to devise a way to reuse parsing code in derived actions and substitutions.

hidmic commented:
> See if we can write a decorator that infers the parsing function from the constructor typing annotations.
That is a really tricky tasks (nice to have).

Yeah, that *is* tricky. I'll try to take a stab at it at some point, should be possible though.

ivanpauno commented on launch_frontend/launch_frontend/action_parse_methods.py:
```diff
@@ -0,0 +1,117 @@
+    return launch.actions.ExecuteProcess(
+        cmd=cmd_list,
+        **kwargs
+    )
```
If we finally modify the `expose_*` decorators to automatically infer the parsing function from the constructor, this will automatically be solved (i.e.: the decorator should not only check the type annotation of the constructor arguments, but it should also check the derived class constructor arguments).
For the moment, I will let this as-is.

ivanpauno replied:
Finally, I have corrected it!

hidmic replied:
> Finally, I have corrected it!

Awesome! And yes, automatic inference may reduce the need for this feature somewhat but we don't plan to work on it soon and we don't know if it's feasible for every use case.

ivanpauno commented on launch_frontend/launch_frontend/type_utils.py:
```diff
@@ -0,0 +1,155 @@
+from typing import Union
+
+
+types_guess__ = (
```
These aren't exactly the supported types, but how `types='guess'` works.
I would let it as-is.

ivanpauno commented on launch_frontend/launch_frontend/type_utils.py:
```diff
@@ -0,0 +1,155 @@
+        if type_obj is bool:
+            def type_obj(x):
+                if x.lower() in ('true', 'false'):
+                    return x.lower() == 'true'
```
1 and 0 too

ivanpauno commented on launch_frontend/launch_frontend/type_utils.py:
```diff
@@ -0,0 +1,155 @@
+        type_obj, is_list = extract_type(x)
+        if type_obj is bool:
+            def type_obj(x):
+                if x.lower() in ('true', 'false'):
```
`myfalsething` raises `ValueError` and doesn't qualify as `False`.

ivanpauno commented on launch_frontend/test/launch_frontend/test_substitutions.py:
```diff
@@ -0,0 +1,136 @@
+def test_combining_quotes_nested_substitution():
+    subst = default_parse_substitution(
+        '$(env "asd_bsd_qsd_$(test \'asd_bds\')" \'$(env DEFAULT)_qsd\')'
+    )
```
@hidmic I added this new test case. For having a more uniform behavior, I think this should look like:
`'$(env "asd_bsd_qsd_$(test \\\'asd_bds\\\')" \'$(env DEFAULT)_qsd\')'`
(this is not a raw a string, I'm escaping both the \ and ')

For that, I would change the following in the grammar:
```
SINGLE_QUOTED_STRING: (/[^'\\"$]|\$(?=!\()|(?<=\\)\$/ | "\\\"" | "\\'" | "\\")+
SINGLE_QUOTED_RSTRING: (/[^ '\\"$\(\)]|(?<=\\)\(|(?<=\\)\)|\$(?=!\()|(?<=\\)\$/ | "\\\"" | "\\'" | "\\")+

DOUBLE_QUOTED_STRING: (/[^"\\'$]|\$(?=!\()|(?<=\\)\$/ | "\\\"" | "\\'" | "\\")+
DOUBLE_QUOTED_RSTRING: (/[^ "\\'$\(\)]|(?<=\\)\(|(?<=\\)\)|\$(?=!\()|(?<=\\)\$/ | "\\\"" | "\\'" | "\\")+
```

Basically, I'm stopping allowing any kind of quote inside `SINGLE_QUOTED_*` and `DOUBLE_QUOTED_*`, and only allowing it if it is escaped.

Do you agree with that change?

hidmic replied:
I'm fine with it, though it may be counter intuitive for some accustomed to combine quotes. If we do it, we should clearly document it somewhere.  @wjwwood any thoughts on this i.e. on forcing quote escaping everywhere if one wants to keep it raw?

ivanpauno commented:
Rebased with master again.

hidmic commented on launch_frontend/launch_frontend/parser.py:
```diff
@@ -0,0 +1,113 @@
+        """Load parser extension, in order to get all the exposed substitutions and actions."""
+        if cls.extensions_loaded is False:
+            for entry_point in iter_entry_points('launch_frontend.parser_extension'):
+                entry_point.load()
```
@ivanpauno meta: we can leave it as is for a first iteration, but I'm not entirely convinced that working with importation side effects is in our best interest in the long run. I wonder if having some sort of index at each package root module that its classes can expose/register themselves into would be cleaner than a having a global one. Maybe @wjwwood can weigh in too on this.

hidmic replied:
It's not so much about performance (though I had not event considered that, thanks for the reference) but the way actions, substitutions and parsers are discovered. It's currently loading a global collection inside `launch` upon entry point loading. 

I was wondering if it wouldn't make sense to move away from that and make exposure somewhat more explicit, and maybe introducing namespaces in the process (which we do not have, we just fail upon tag name collision). Anyways, just a thought. No need to address this now, really.

ivanpauno commented on launch_frontend/launch_frontend/action_parse_methods.py:
```diff
@@ -0,0 +1,131 @@
+        # argument is supposed to be a dictionary with `SomeSubstitutionType`
+        # keys, but `SomeSubstitutionType` is not always hashable.
+        # Proposed `additional_env` type:
+        #   Iterable[Tuple[SomeSubstitutionType, SomeSubstitutionsType]]
```
See https://github.com/ros2/launch/issues/260.

ivanpauno commented on launch_frontend/launch_frontend/entity.py:
```diff
@@ -0,0 +1,97 @@
+            - 'list[Entity]'
+
+        'guess' work in the same way as:
+            ('int', 'float', 'bool', 'list[int]', 'list[float]', 'list[bool]', 'list[str]', 'str')
```
Because the type that is specified most of the times is `str` (substitutions are `str`).

hidmic replied:
Same, making `str` the default.

ivanpauno commented on launch_frontend/launch_frontend/parse_substitution.py:
```diff
@@ -0,0 +1,108 @@
+        assert isinstance(name, Token)
+        assert name.type == 'IDENTIFIER'
+        # TODO(hidmic): Lookup and instantiate Substitution
+        if name.value not in substitution_parse_methods:
```
Could you expand your idea? I don't understand the benefits of using an `instantiate_substitution` function (it will finally access the internal data).

hidmic replied:
It will access data structures eventually, yes, but you stabilize an API in case anything changes in the future (e.g. how we instantiate, how lookup, what the data structure is, etc.).

hidmic replied:
Still applies.

hidmic replied:
I'm fine with bringing another reviewer but I still think this applies.

ivanpauno commented on launch_frontend/package.xml:
```diff
@@ -0,0 +1,20 @@
+<package format="2">
+  <name>launch_frontend</name>
+  <version>0.8.1</version>
+  <description>The ROS launch frontend.</description>
```
https://github.com/ros2/launch/blob/188c773482a255088b084f7bc18c6dd5b9e05f31/launch/package.xml#L4-L6

hidmic replied:
That's simply wrong :)

ivanpauno commented on launch_frontend/launch_frontend/parser.py:
```diff
@@ -0,0 +1,113 @@
+        return default_parse_substitution(value)
+
+    @classmethod
+    def parse_description(cls, entity: Entity) -> launch.LaunchDescription:
```
`parse_description`, `parse_action` should. `cls` is finally passed to the action parsing methods as `parser` argument. The parsing methods could call `parse.substitution`, which can be overridden by the derived class.

hidmic replied:
Right, my point being that these should be instance methods.

ivanpauno replied:
The `Parse` instances don't have a state now, but it could have it in the future, so I think it is a good idea. I will change `parse_action`, `parse_substitution`, `parse_description` to instance methods.
`load`, `load_parser_extensions` and `load_parser_implementations` will be continue being class methods.
I will stop returning an object of `Parse` type in the `load` method (ie: a subclass of `Parse`) , and start returning an instance of it.
Note: I will also rename `load_parser_extensions` as `load_launch_extensions`.

ivanpauno commented on launch_frontend/launch_frontend/parser.py:
```diff
@@ -0,0 +1,113 @@
+    @classmethod
+    def load(
+        cls,
+        file: Union[str, io.TextIOBase],
```
`file` isn't a builtin in python3 (it was in python 2).

hidmic replied:
Ok, fair enough.

ivanpauno commented on launch_frontend/launch_frontend/type_utils.py:
```diff
@@ -0,0 +1,155 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
The problem is with `'list[int]'`, etc. Using `typing` objects doesn't help much in the implementation (i.e.: I don't have an easy way to extract if it was a list, and what is the item type).

hidmic replied:
Right, but you can check for equality just like you do with strings. And then you don't have to do substring checks.

hidmic replied:
I still wonder if it wouldn't be better to use `typing` objects instead of plain `str`ings.

hidmic replied:
So, between using `typing` classes and using a stringification of their names, I'm slightly inclined towards the first one. Collection type information is readily available and the API looks cleaner that way IMO. No need to do substring checks anymore.

ivanpauno commented:
@hidmic Addressed most of your comments in https://github.com/ros2/launch/pull/226/commits/61093a53a83a391043d0ac252a3445ea522cbca7.
Also, rebased with master.

hidmic commented on launch_frontend/launch_frontend/action_parse_methods.py:
```diff
@@ -0,0 +1,126 @@
+    # `args` is supposed to be a list separated with ' '.
+    # All the found `TextSubstitution` items are split and
+    # added to the list again as a `TextSubstitution`.
+    # TODO(ivanpauno): Change `ExecuteProcess` api from accepting
```
@ivanpauno same as the other TODO, I think we should file this or it'll get lost.

hidmic commented on launch_frontend/launch_frontend/entity.py:
```diff
@@ -0,0 +1,96 @@
+        List['Entity']
+    ]]:
+        """
+        Access an element of the entity.
```
@ivanpauno `element` -> `attribute`?

hidmic commented on launch_frontend/launch_frontend/entity.py:
```diff
@@ -0,0 +1,96 @@
+            It returns `None` instead. Defaults to `False`
+        :raises `AttributeError`: Attribute not found. Only possible if `optional` is `False`
+        :raises `TypeError`: Attribute found but it is not of the correct type.
+            Only happens in froentends that do type checking
```
@ivanpauno consider `froentends` -> `frontend implementations`.

hidmic commented on launch_frontend/launch_frontend/entity.py:
```diff
@@ -0,0 +1,96 @@
+        :raises `TypeError`: Attribute found but it is not of the correct type.
+            Only happens in froentends that do type checking
+        :raises `ValueError`: Attribute found but can't be coerced to one of the types.
+            Only happens in froentends that do type coercion
```
@ivanpauno consider `froentends` -> `frontend implementations`.

hidmic commented on launch/launch/launch_description_sources/any_launch_description_source.py:
```diff
@@ -0,0 +1,69 @@
+    """
+    Encapsulation of a launch file, which can be loaded during launch.
+
+    This try to load the file as a python launch file, or as a frontend launch file.
```
@ivanpauno consider:
```suggestion
    This launch description source will attempt to load the file at the given location as a python launch file first, and as a declarative (markup based) launch file if the former fails.

hidmic commented on launch/launch/launch_description_sources/any_launch_description_source.py:
```diff
@@ -0,0 +1,69 @@
+        The path can be made up of Substitution instances which are expanded
+        when :py:meth:`get_launch_description()` is called.
+
+        :param launch_file_path: the path to the launch file
```
@ivanpauno missing period at the end of the sentence.

hidmic commented on launch/launch/launch_description_sources/any_launch_description_source.py:
```diff
@@ -0,0 +1,69 @@
+        directory will be wherever the launch file was run from, which might
+        change depending on the situation.
+        The path can be made up of Substitution instances which are expanded
+        when :py:meth:`get_launch_description()` is called.
```
@ivanpauno consider moving this all these details to the parameter documentation itself.

hidmic commented on launch/launch/launch_description_sources/any_launch_description_source.py:
```diff
@@ -0,0 +1,69 @@
+        The given file path should be to a launch file of any style.
+        The path should probably be absolute, since the current working
+        directory will be wherever the launch file was run from, which might
+        change depending on the situation.
```
@ivanpauno hmm, why not simply stating the outcome of passing a relative path?

hidmic commented on launch/launch/launch_description_sources/frontend_launch_description_source.py:
```diff
@@ -0,0 +1,57 @@
+        directory will be wherever the launch file was run from, which might
+        change depending on the situation.
+        The path can be made up of Substitution instances which are expanded
+        when :py:meth:`get_launch_description()` is called.
```
@ivanpauno same as above.

hidmic commented on launch/launch/launch_description_sources/xml_launch_description_source.py:
```diff
@@ -0,0 +1,46 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
@ivanpauno hmm, shouldn't this source live in `launch_xml`?

hidmic commented on launch/launch/launch_description_sources/yaml_launch_description_source.py:
```diff
@@ -0,0 +1,46 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
@ivanpauno same, shouldn't this source live in `launch_yaml`?

hidmic commented on launch/launch/launch_description_sources/yaml_launch_description_source.py:
```diff
@@ -0,0 +1,46 @@
+        directory will be wherever the launch file was run from, which might
+        change depending on the situation.
+        The path can be made up of Substitution instances which are expanded
+        when :py:meth:`get_launch_description()` is called.
```
@ivanpauno same on documentation.

hidmic commented on launch_frontend/launch_frontend/action_parse_methods.py:
```diff
@@ -0,0 +1,155 @@
+    """
+    Parse action.
+
+    This is only intented for code reusage, and it's not exposed.
```
@ivanpauno consider:
```suggestion
    This is only intended for code reuse, and it's not exposed.
```

hidmic commented on launch_frontend/launch_frontend/action_parse_methods.py:
```diff
@@ -0,0 +1,155 @@
+            )
+            for e in args
+        ]
+        kwargs['launch_arguments'] = args
```
@ivanpauno consider collapsing above two lines in one:

```python
        kwargs['launch_arguments'] = [
            (
                parser.parse_substitution(e.get_attr('name')),
                parser.parse_substitution(e.get_attr('value'))
            )
            for e in args
        ]
```

hidmic commented on launch_xml/test/launch_xml/test_executable.py:
```diff
@@ -0,0 +1,58 @@
+    value = value[0].perform(None)
+    assert(key == 'var')
+    assert(value == '1')
+    # assert(executable.prefix[0].perform(None) == 'time')
```
@ivanpauno remove?

hidmic commented on launch_xml/test/launch_xml/test_include.py:
```diff
@@ -0,0 +1,44 @@
+    include = ld.entities[0]
+    assert isinstance(include, IncludeLaunchDescription)
+    assert isinstance(include.launch_description_source, AnyLaunchDescriptionSource)
+    # TODO(ivanpauno): Load something really
```
@ivanpauno do you think we can do it for this PR?

wjwwood replied:
It would be good to see it work.

hidmic commented on launch_yaml/launch_yaml/parser.py:
```diff
@@ -0,0 +1,43 @@
+        """Load a YAML launch file."""
+        if isinstance(stream, str):
+            stream = open(stream, 'r')
+        """Return entity loaded with markup file."""
```
@ivanpauno hmm, a leftover?

hidmic commented on launch/launch/__init__.py:
```diff
@@ -38,7 +40,9 @@
     'actions',
     'conditions',
     'events',
+    'launch_frontend',
```
@ivanpauno can't we just call it `frontend`? To go from `launch.launch_frontend` to `launch.frontend`.

hidmic commented on launch/launch/actions/execute_process.py:
```diff
@@ -225,6 +230,71 @@ def __init__(
+            args = []
+        cmd_list.extend(args)
+        kwargs['cmd'] = cmd_list
+        _, action_kwargs = super(ExecuteProcess, ExecuteProcess).parse(entity, parser)
```
@ivanpauno consider moving this to the top instead of default initializing `kwargs`.

hidmic commented on launch/launch/actions/execute_process.py:
```diff
@@ -225,6 +230,71 @@ def __init__(
+                    text = arg.text
+                    text = shlex.split(text)
+                    text = [TextSubstitution(text=item) for item in text]
+                    new_args.extend(text)
```
@ivanpauno can't we collapse this as:

```python
new_args.extend([TextSubstitution(text=token) for token in shlex.split(args.text)])
```

ivanpauno replied:
is not the same

hidmic commented on launch/launch/actions/execute_process.py:
```diff
@@ -225,6 +230,71 @@ def __init__(
+    def parse(
+        entity: Entity,
+        parser: Parser,
+        optional_cmd: bool = False
```
@ivanpauno meta: let's revisit this mechanism for reuse after this PR.

hidmic commented on launch/launch/actions/group_action.py:
```diff
@@ -54,6 +58,16 @@ def __init__(
+    @staticmethod
+    def parse(entity: Entity, parser: Parser):
+        """Return `GroupAction` action and kwargs for constructing it."""
+        _, kwargs = super(GroupAction, GroupAction).parse(entity, parser)
```
@ivanpauno I think I've seen enough of this idiom to think it'd be better to use class methods, which then allows:
```python
super(GroupAction, cls).parse(entity, parser)
```
which IMHO is slightly less terse.

ivanpauno replied:
Yeah, I don't know why I didn't use classmethods.
You can use `super()` without any arguments in them, which looks better.

hidmic commented on launch/launch/actions/include_launch_description.py:
```diff
@@ -69,6 +74,26 @@ def __init__(
+            IncludeLaunchDescription,
+            IncludeLaunchDescription
+        ).parse(entity, parser)
+        file = parser.parse_substitution(entity.get_attr('file'))
```
@ivanpauno nit: I know I brought this up before, and that `file` is no longer a Python keyword, but I can count 3 syntax highlighters that think otherwise. Can we rename it?

hidmic commented on launch/launch/launch_description_sources/any_launch_description_source.py:
```diff
@@ -0,0 +1,66 @@
+
+    This launch description source will attempt to load the file at the given location as a python
+    launch file first, and as a declarative (markup based) launch file if the former fails.
+    It is recommended to use a specific `LaunchDescriptionSource` subclasses when possible.
```
@ivanpauno consider
```suggestion
    It is recommended to use specific `LaunchDescriptionSource` subclasses when possible.
```

hidmic commented on launch/launch/launch_description_sources/any_launch_description_source.py:
```diff
@@ -0,0 +1,66 @@
+        except RuntimeError:
+            pass
+        if launch_description is None:
+            raise RuntimeError('Can not load launch file')
```
@ivanpauno consider 
```suggestion
            raise RuntimeError('Cannot load launch file')
```

hidmic commented on launch/launch/launch_frontend/entity.py:
```diff
@@ -0,0 +1,104 @@
+        Types that can not be combined with the others:
+            - `List[Entity]`
+
+        `types = None` work in the same way as:
```
@ivanpauno consider
```suggestion
        `types=None` implies `types=(...)`
```

hidmic commented on launch/launch/launch_frontend/entity.py:
```diff
@@ -0,0 +1,104 @@
+
+        `types = None` work in the same way as:
+            `(int, float, bool, List[int], List[float], List[bool], List[str], str)`
+        `List[Entity]` will return a list of more entities.
```
@ivanpauno consider
```suggestion
        `List[Entity]` will return the list of entities that are identified by the given attribute.
```

hidmic commented on launch/launch/launch_frontend/entity.py:
```diff
@@ -0,0 +1,104 @@
+        See the frontend documentation to see how `list` and `List[Entity]` look like for each
+        frontend implementation.
+
+        If `optional` argument is `True`, will return `None` instead of raising `AttributeError`.
```
@ivanpauno consider
```suggestion
        If `optional` is `True` and the attribute cannot be found, `None` will be returned instead of raising `AttributeError`.
```

hidmic commented on launch/launch/substitutions/environment_variable.py:
```diff
@@ -43,6 +46,17 @@ def __init__(
+        if not data or len(data) > 2:
+            raise TypeError('env substitution expects 1 or 2 arguments')
+        kwargs = {}
+        kwargs['name'] = data[0]
```
@ivanpauno nit: you can also do

```python
kwargs = {'name': data[0]}
```

hidmic commented on launch/launch/substitutions/find_executable.py:
```diff
@@ -39,6 +42,15 @@ def __init__(self, *, name: SomeSubstitutionsType) -> None:
+    @staticmethod
+    def parse(data: Iterable[SomeSubstitutionsType]):
+        """Parse `FindExecutable` substitution."""
+        if not data or len(data) > 1:
```
@ivanpauno consider
```suggestion
        if len(data) != 1:
```

hidmic commented on launch/launch/substitutions/find_executable.py:
```diff
@@ -39,6 +42,15 @@ def __init__(self, *, name: SomeSubstitutionsType) -> None:
+            raise AttributeError('find-exec substitution expects 1 argument')
+        kwargs = {}
+        kwargs['name'] = data[0]
+        return FindExecutable, kwargs
```
@ivanpauno nit: you can also just do

```python
return FindExecutable, {'name': data[0]}
```

hidmic commented on launch/launch/substitutions/launch_configuration.py:
```diff
@@ -62,6 +64,17 @@ def __init__(
+    @staticmethod
+    def parse(data: Iterable[SomeSubstitutionsType]):
+        """Parse `FindExecutable` substitution."""
+        if not data or len(data) > 2:
```
@ivanpauno consider
```suggestion
        if len(data) < 1 or len(data) > 2:
```

hidmic commented on launch_xml/launch_xml/entity.py:
```diff
@@ -0,0 +1,117 @@
+            else:
+                return None
+        try:
+            value = get_typed_value(value, types)
```
@ivanpauno hmm why are we mixing YAML and XML here?

ivanpauno replied:
I'm only doing the conversions from string to int/float/bool like yaml, and allowing the yaml ways of escaping them. If not it was imposible to specify an string param of value "1" (now you can do that passing "'1'".

Maybe this is not the best solution, if you have any other ideas please tell me.

Maybe, one option is to add an optional type field in `param`. If you don't use it, the automatic type conversions are done as before ("1" is an int "'1'" doesn't escape the quotes, you see them).
If you want a string param containing 1 you could do:
```xml
<param name="asd" value="1" type="str"/>
```

I like this last option. It should be possible to use the `type` attribute in yaml too, but it hasn't much sense for the user.

hidmic replied:
My point is that with `get_typed_value()`'s current implementation, you can pass a list of floats using `[1, 2, 3]`. And then we also have value separators attributes.

ivanpauno replied:
I restricted that conversion in https://github.com/ros2/launch/pull/226/commits/18cf4a06acaa0b5002e7b5c5f4fe90e03b08651d.
For parameters inside substitutions, you can still pass a list like yaml (becuase of https://github.com/ros2/launch_ros/pull/31).
So this example is still working:
https://github.com/ros2/launch_ros/blob/713fe0034d89c412dd0efbd327ce31aecdb359c0/test_launch_ros/test/test_launch_ros/launch_frontend/test_node_frontend.py#L31L50

hidmic commented on launch_xml/test/launch_xml/executable.xml:
```diff
@@ -0,0 +1,5 @@
+<launch>
+    <executable cmd="ls" cwd="/" name="my_ls" args="-l -a -s" shell="true" output='log' launch-prefix='$(env LAUNCH_PREFIX)'>
```
@ivanpauno nit: I wonder if just collapsing `args` into `cmd` wouldn't be simpler: `ls -l -a -s`

ivanpauno commented:
Finally, CI (only fastrtps, up to launch launch_ros launch_yaml launch_xml test_launch_ros):
* Linux [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7492)](https://ci.ros2.org/job/ci_linux/7492/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3630)](http://ci.ros2.org/job/ci_linux-aarch64/3630/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6142)](http://ci.ros2.org/job/ci_osx/6142/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7322)](http://ci.ros2.org/job/ci_windows/7322/)

ivanpauno commented:
Again, now that CI problems have been solved:
* Linux [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7505)](https://ci.ros2.org/job/ci_linux/7505/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3645)](http://ci.ros2.org/job/ci_linux-aarch64/3645/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6150)](http://ci.ros2.org/job/ci_osx/6150/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7331)](http://ci.ros2.org/job/ci_windows/7331/)

ivanpauno commented:
Checking if windows is happy after https://github.com/ros2/launch/pull/226/commits/f15b0f21739d9c2b4514a658e73287238d7642c9:
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7332)](http://ci.ros2.org/job/ci_windows/7332/)

ivanpauno commented:
CI ...
* Linux [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7512)](https://ci.ros2.org/job/ci_linux/7512/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3648)](http://ci.ros2.org/job/ci_linux-aarch64/3648/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6152)](http://ci.ros2.org/job/ci_osx/6152/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7335)](http://ci.ros2.org/job/ci_windows/7335/)

hidmic commented on launch/launch/frontend/entity.py:
```diff
@@ -0,0 +1,87 @@
+            - a non-uniform list.
+                Can specify some of the scalar types for its members, e.g.: List[Union[int, str]].
+                `list` or `List`, means that any of the scalar types are allowed.
+            - an union of the above.
```
@ivanpauno consider: 

```
The allowed types are:
  - a scalar type i.e. `str`, `int`, `float`, `bool`;
  - a uniform list i.e `List[str]`, `List[int]`, `List[float]`, `List[bool]`;
  - a non-uniform list of known scalar types e.g. `List[Union[int, str]]`;
  - a non-uniform list of any scalar type i.e. `list` or `List`;
  - a `Union` of any of the above.
```

hidmic commented on launch/launch/frontend/entity.py:
```diff
@@ -0,0 +1,87 @@
+        `types = None` works in the same way as:
+            `Union[int, float, bool, list, str]`
+
+        See the frontend documentation to see how `list` and `List[Entity]` look like for each
```
@ivanpauno consider
```suggestion
        Check the documentation of each specific frontend implementation to see how `list` and `List[Entity]` look like.
```

hidmic commented on launch/launch/frontend/type_utils.py:
```diff
@@ -0,0 +1,254 @@
+
+
+def check_is_list(data_type: Any) -> bool:
+    """Check if `data_type` was origined from `typing.List`."""
```
@ivanpauno consider
```suggestion
    """Check if `data_type` is based on a `typing.List`."""
```

hidmic commented on launch/launch/frontend/type_utils.py:
```diff
@@ -0,0 +1,254 @@
+
+
+def check_is_union(data_type: Any) -> bool:
+    """Check if `data_type` was origined from `typing.Union`."""
```
@ivanpauno consider
```suggestion
    """Check if `data_type` is based on a `typing.Union`."""
```

hidmic commented on launch/launch/frontend/type_utils.py:
```diff
@@ -0,0 +1,254 @@
+
+def check_is_list_entity(data_type: Any) -> bool:
+    """
+    Return `True` if `data_type` is `List[launch.frontend.Entity]`.
```
@ivanpauno consider
```suggestion
    Check if `data_type` is a `typing.List` with elements of `launch.frontend.Entity` type or derived types.
```

hidmic commented on launch/launch/frontend/type_utils.py:
```diff
@@ -0,0 +1,254 @@
+
+
+def get_tuple_of_types(data_type: Any) -> Tuple:
+    """Convert typing.Union to tuple of types. If not, return `(data_type,)`."""
```
@ivanpauno consider
```suggestion
    """
    Normalize `data_type` to a tuple of types. 
    
    If `data_type` is based on a `typing.Union`, return union types. 
    Otherwise, return a `(data_type,)` tuple.
    """
```

hidmic commented on launch/launch/frontend/type_utils.py:
```diff
@@ -0,0 +1,254 @@
+
+
+def check_valid_scalar_type(data_type: Any) -> bool:
+    """Check if it is a valid scalar type."""
```
@ivanpauno consider
```suggestion
    """Check if `data_type` is a valid scalar type."""
```

hidmic commented on launch/launch/frontend/type_utils.py:
```diff
@@ -0,0 +1,254 @@
+        - a uniform list: `List[str]`, `List[int]`, `List[float]`, `List[bool]`
+        - a non-uniform list.
+            Can specify some of the scalar types for its members, e.g.: List[Union[int, str]].
+            `list` or `List`, means that any of the scalar types are allowed.
```
@ivanpauno see similar docstring suggestion above.

hidmic commented on launch/launch/frontend/type_utils.py:
```diff
@@ -0,0 +1,254 @@
+    keys = {
+        int: 0,
+        float: 1,
+        bool: 2,
```
@ivanpauno meta: I wonder if `bool` should come before `int`

ivanpauno replied:
No, if not `0` will be converted to bool. If both are possible, it should be an int.
And maybe we should just delete that option. So the rules do not overlap.

hidmic replied:
> If both are possible, it should be an int.

Ok, fair enough, Python coercion will do its job.

hidmic commented on launch_xml/README.md:
````diff
@@ -0,0 +1,98 @@
+For handling lists, the `*-sep` attribute is used. e.g.:
+
+```xml
+<tag value="2,3,4" sth-sep=","/>
````
@ivanpauno shouldn't it be `value-sep` instead?

ivanpauno commented:
@hidmic I've addressed all the comments in both PRs. I will run an extra CI after an approval.

ivanpauno commented:
Luckly, last CI:
* Linux [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7531)](https://ci.ros2.org/job/ci_linux/7531/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3668)](http://ci.ros2.org/job/ci_linux-aarch64/3668/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6168)](http://ci.ros2.org/job/ci_osx/6168/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7360)](http://ci.ros2.org/job/ci_windows/7360/)
